### PR TITLE
Add CUDA graph capture/replay for qwen 3.5 moe decode method

### PR DESCRIPTION
**Pull request by Gasoonjia**
**Problem**: In the ExecuTorch CUDA backend, each decode step re-launches all GPU kernels individually, incurring significant CPU-side kernel launch overhead. For autoregressive LLM decoding (single-token steps), this launch overhead dominates end-to-end latency.

**Solution**: Added CUDA graph support to the CUDA backend. When enabled for a method (e.g., decode), the backend:

- Uses the first three decode iteration as  warmup to stabilize GPU state (lazy allocations, cuBLAS handles, kernel JIT)
- Captures the 4th execution into a CUDA graph
- Replays the captured graph on all subsequent decode calls, eliminating per-step kernel launch overhead

warm up phase vs cuda graph replay phase: 

```
Phase                            Per-token time            Throughput
*Warmup* (3 tokens, no graph)  ~11.5 ms                  ~87 tok/s
*Capture* (1 token)            ~21.4 ms (one-time cost)  --
*Replay* (508 tokens)          *7.60 ms*               *~132 tok/s*
```

### CODE REVIEW

File: backends/cuda/runtime/cuda_backend.cpp [hunk 1]
```diff
@@ -548,6 +638,41 @@ class ET_EXPERIMENTAL CudaBackend final
...
+            merr == cudaSuccess, Internal,
+            "cudaMalloc for static input %zu failed: %s",
+            i, cudaGetErrorString(merr));
+
+        cudaMemcpy(
+            static_ptr, cpu_tensor->const_data_ptr(),
+            nbytes, cudaMemcpyHostToDevice);
+
+        handle->static_input_ptrs.push_back(static_ptr);
+        handle->static_input_sizes.push_back(sizes_vec);
```
Comment (digantdesai) [inline backends/cuda/runtime/cuda_backend.cpp]: Does this mean a method must have all static shapes for its intermediate tensors?

Comment (Gasoonjia) [reply backends/cuda/runtime/cuda_backend.cpp]: yes the cuda graph can only support a static pipeline.

**Comment by Gasoonjia** [comment]
@claude review this PR

**Comment by Gasoonjia** [comment]
have solved all claude issue

File: .ci/scripts/test_model_e2e.sh [hunk 2]
```diff
@@ -354,7 +354,9 @@ EOF
     fi
     ;;
   qwen3_5_moe)
-    RUNNER_ARGS="$RUNNER_ARGS --tokenizer_path ${MODEL_DIR}/$TOKENIZER_FILE --prompt 'What is the capital of France?' --max_new_tokens 128 --temperature 0"
+    RUNNER_ARGS="$RUNNER_ARGS --tokenizer_path ${MODEL_DIR}/$TOKENIZER_FILE --prompt 'What is the capital of France?' --max_new_tokens 128 --temperature 0 --cuda_graph"
+    # CUDA graph capture requires cudaMallocAsync backend for stream-ordered allocations
+    export PYTORCH_CUDA_ALLOC_CONF=backend:cudaMallocAsync
```
Comment (digantdesai) [inline .ci/scripts/test_model_e2e.sh]: why does it have to be specified here, can we enable it internally when user specifies `--cuda_graph`?

**Comment by Gasoonjia** [comment]
removed backend:cudaMallocAsync variable.

File: .ci/scripts/test_model_e2e.sh [hunk 3]
```diff
@@ -354,7 +354,7 @@ EOF
     fi
     ;;
   qwen3_5_moe)
-    RUNNER_ARGS="$RUNNER_ARGS --tokenizer_path ${MODEL_DIR}/$TOKENIZER_FILE --prompt 'What is the capital of France?' --max_new_tokens 128 --temperature 0"
+    RUNNER_ARGS="$RUNNER_ARGS --tokenizer_path ${MODEL_DIR}/$TOKENIZER_FILE --prompt 'What is the capital of France?' --max_new_tokens 128 --temperature 0 --cuda_graph"
```
Comment (digantdesai) [inline .ci/scripts/test_model_e2e.sh:357]: Do we also want to test w/o cuda_graph?

File: backends/cuda/runtime/cuda_backend.cpp [hunk 4]
```diff
@@ -80,6 +81,8 @@ namespace {
 constexpr char kSkipCopyOutputToCpuForMethod[] =
     "skip_copy_output_to_cpu_for_method";
 constexpr char kUseSharedCudaStream[] = "use_shared_cuda_stream";
+constexpr char kEnableCudaGraphForMethod[] = "enable_cuda_graph_for_method";
+constexpr int kCudaGraphWarmupSteps = 3;
```
Comment (digantdesai) [inline backends/cuda/runtime/cuda_backend.cpp:85]: make it user configurable?

File: backends/cuda/runtime/cuda_backend.cpp [hunk 5]
```diff
@@ -667,6 +801,87 @@ class ET_EXPERIMENTAL CudaBackend final
...
+          gerr == cudaSuccess,
+          Internal,
+          "cudaGraphLaunch (first replay) failed: %s",
+          cudaGetErrorString(gerr));
+
+      // Copy capture-step outputs to CPU
+      const bool copy_outputs =
+          !should_skip_copy_for_method(handle->method_name);
+      if (copy_outputs) {
+        for (size_t i = 0; i < n_outputs; i++) {
```
Comment (digantdesai) [inline backends/cuda/runtime/cuda_backend.cpp:849]: can we refactor so that this code is shared for regular replay run and capture run?

File: backends/cuda/runtime/cuda_delegate_handle.h [hunk 6]
```diff
@@ -38,6 +39,93 @@ inline std::shared_ptr<cudaStream_t> create_cuda_stream() {
   return std::shared_ptr<cudaStream_t>(
       new cudaStream_t(stream), CudaStreamDeleter());
 }
+
+enum class CudaGraphPhase {
```
Comment (digantdesai) [inline backends/cuda/runtime/cuda_delegate_handle.h:62]: add comments on what each does

**Review by digantdesai — APPROVED** [review]
Thanks, do we want to include some stats for cuda_graph in terms of how much overhead it reduces by comparing 1st or 2nd run vs. nth run post enablement phase?

File: .ci/scripts/test_model_e2e.sh (hunk 3 above)
Comment (Gasoonjia) [reply .ci/scripts/test_model_e2e.sh:357]: i don;t think we need test w/o cuda_graph; i don't think in moe case cuda graph decoder will have any negative impact to the final state.

File: backends/cuda/runtime/cuda_backend.cpp (hunk 5 above)
Comment (Gasoonjia) [reply backends/cuda/runtime/cuda_backend.cpp:849]: no we can't; for warm up phase we should delete the gpu tensor for memory managerment; however for capture and replay we can;t modify the input and output tensor managerment due to cuda graph limitation.

Furthermore after device support stack lands, we should delegate all tensor managerment into ET graph memory planing, and all verbose code here will be gone,

File: backends/cuda/runtime/cuda_backend.cpp (hunk 4 above)
Comment (Gasoonjia) [reply backends/cuda/runtime/cuda_backend.cpp:85]: We can keep as it is for now. We can make it configurable in the future if we actialy have that requirement.

**Comment by Gasoonjia** [comment]
> Thanks, do we want to include some stats for cuda_graph in terms of how much overhead it reduces by comparing 1st or 2nd run vs. nth run post enablement phase?

Thanks! Have update the top comment, added the comparison between warm up runs and cuda reply runs, and add comments for CudaGraphPhase enum as well.
